builder/slide_and_fold_storage.cc

```diff
@@ -514,15 +514,15 @@ class slide_and_fold : public stmt_mutator {
     if (bounds) {
       define_undef_bounds(*bounds, op->sym);
       substitute_bounds(*bounds, current_buffer_bounds());
-      // Now do the reverse substitution, because the updated bounds can be used in other
-      // bounds.
-      substitute_bounds(current_buffer_bounds(), op->sym, *bounds);
-
       // This simplify can be heavy, but is really useful in reducing the size of the
       // expressions.
       for (auto& b : *bounds) {
         b = simplify(b);
       }
+
+      // Now do the reverse substitution, because the updated bounds can be used in other
+      // bounds.
+      substitute_bounds(current_buffer_bounds(), op->sym, *bounds);
```

Comment:
Should we simplify again after this?

Reply:
Maybe it would help, but this is different: we'd have to simplify *every* bounds expression in scope.

I think we need to not even do the other simplification eventually, but as long as we're doing it, we might as well do it before using the unsimplified versions (this PR).